Localization with Unicode CLDR and overrideable strings (#2894)
microsoft/BotFramework-WebChat: docs/LOCALIZATION.md

# Localization

> Starting from Web Chat 4.8, a dedicated team at Microsoft will start localizing our UI for 44 languages.

Reviewer: ```suggestion
Beginning in Web Chat 4.8, this project shifted from community-provided localization to supporting most languages through a dedicated Microsoft team. As of 2020.02.14, 44 languages are officially maintained by Microsoft. Other languages will continue to be community-supported and -driven.
```

> ## Adding a new language

Reviewer: Link to these directions within `CONTRIBUTING.md`

> To add a new language, please update following files:

Reviewer: ```suggestion
To add a new language, please update the following files:
```

> -  Update `/packages/component/src/Localization/overrides.json` and add the language according to Unicode CLDR.

Reviewer: add link to unicode site on last two words?
http://cldr.unicode.org/

Reviewer: or link to the package below

   -  `GLOBALIZE_LANGUAGE` is the language code used for referencing Unicode CLDR
      -  To see the supported list of Unicode CLDR, look under `/packages/component/node_modules/cldr-data/main/`.
   -  (Optional) `TEXT_TO_SPEECH` is used to indicate the language code for speech.

>       -  Some locales has different oral but share same written language with other locales. For example, written language in Hong Kong and Taiwan are Traditional Chinese, while oral languages are Cantonese and Taiwanese Mandarin respectively.

Reviewer: ```suggestion
      -  Some locales have different oral languages, but share same written language with other locales. For example, the written language in both Hong Kong and Taiwan are Traditional Chinese, while oral languages are Cantonese and Taiwanese Mandarin respectively.
```

>       -  The language code used in Web Chat only indicate the written language.

Reviewer: ```suggestion
      -  The language code used in Web Chat only indicates the written language.
```

>       -  Thus, when performing text-to-speech, the language code will be remapped for oral language.

Reviewer: ```suggestion
      -  Thus, when performing text-to-speech, the language code will be remapped to the oral language.
```

>       -  For example, written language in Hong Kong is `zh-Hant-HK`, while the oral language is `zh-HK` (Cantonese). And the written language in Taiwan is `zh-Hant`, while the oral language is `zh-TW` (Taiwanese Mandarin).

Reviewer: ```suggestion
      -  For example, the written language in Hong Kong is `zh-Hant-HK`, while the oral language is `zh-HK` (Cantonese). Meanwhile, the written language in Taiwan is `zh-Hant`, and the oral language is `zh-TW` (Taiwanese Mandarin).
```

>    -  (Optional) `COGNITIVE_SERVICES_*` is used to indicate whether the language supported Cognitive Services Speech Services.

Reviewer: is supported by?

>       -  `COGNITIVE_SERVICES_SPEECH_TO_TEXT` is used to indicate the language is supported by Cognitive Services Speech-to-Text Service.

Reviewer: ```suggestion
      -  `COGNITIVE_SERVICES_SPEECH_TO_TEXT` is used to indicate if the language is supported by Cognitive Services Speech-to-Text Service.
```

>       -  `COGNITIVE_SERVICES_TEXT_TO_SPEECH` is used to indicate the language is supported by Cognitive Services Text-to-Speech Service. If the service support neural voices, set it to `"neural"`, otherwise, `true`

Reviewer: ```suggestion
      -  `COGNITIVE_SERVICES_TEXT_TO_SPEECH` is used to indicate that the language is supported by Cognitive Services Text-to-Speech Service. If the service supports neural voices, set it to `"neural"`; otherwise, `true`
```

>       -  https://docs.microsoft.com/en-us/azure/cognitive-services/speech-service/language-support

Reviewer: Accessibility: add as anchor with description of what the link is

-  Update `/packages/component/src/Utils/normalizeLanguage.js` and add a normalization logic for ISO language.
   -  This is for cleaning up data provided by developers.
   -  For example, if the language starts with `ja`, it will be mapped to `ja-JP`.
-  Add a new language to `/packages/component/src/Localization/`.
   -  Copy `en-US.json` as the base template.

## Updating strings of an existing language

There are two types of supported languages:

-  Maintained by Microsoft
-  Contributed by the community

> For strings that are maintained by Microsoft, please file a bug to us. It will take us time as we need perform additional checks to maintain the quality.

Reviewer: ```suggestion
For strings that are maintained by Microsoft, please file a bug to the Web Chat repo. Validation will be performed on these changes to maintain high quality.
```

> For strings that are contributed by the community, please submit a pull request to us.

Reviewer: ```suggestion
For strings that are contributed by the community, please submit a pull request to the Web Chat repo.
```

## Adding new strings for new UI

> To add new strings to be used in the UI, please update `/packages/component/src/localization/en-US.json`. Our localization team will pick up the new strings and translate them to all other officially supported languages.

Reviewer: ```suggestion
To add new strings to be used in the UI, please update `/packages/component/src/localization/en-US.json`. The localization team will pick up the new strings and translate them to all other officially supported languages.
```

## Overriding localization strings

> Developers can override localization strings in one of two ways. An `overrideLocalizedStrings` prop will be passed and it can either be a plain object or a function.

Reviewer: ```suggestion
Developers can override localization strings in one of two ways. An `overrideLocalizedStrings` prop can be passed either as a plain object or a function in the Web Chat renderer.
```

```js
window.WebChat.renderWebChat(
   {
      locale: 'en-US',

      overrideLocalizedStrings: (strings, language) => ({
         ...strings,
         TEXT_INPUT_PLACEHOLDER: 'What is on your mind?'
      }),

      overrideLocalizedStrings: {
         TEXT_INPUT_PLACEHOLDER: 'What is on your mind?'
      }
   },
   document.getElementById('webchat')
);
```

## Design

### User stories

The localization feature is backed by these user stories:

-  Translated by Microsoft localization team with automated pipeline
-  Overrideable by developers
   -  Change the "Type your message" placeholder text

> 

Reviewer: Is this necessary? I don't think it's relevant to the customer.

Reviewer: If it is, I think it needs more clarity

Author: Updated.

### Translated by Microsoft localization team

> We support 44 languages, including right-to-left languages, translated by Microsoft localization team. Additional languages by the community are always welcomed.

Reviewer: ```suggestion
Web Chat officially supports 44 languages translated by the Microsoft localization team, which includes right-to-left languages.
Additional languages contributed and maintained by the community are always welcome.
```

| Language Code | Language              | Name             |
| ------------- | --------------------- | ---------------- |
| `ar-SA`       | Arabic (Saudi Arabia) | اَلْعَرَبِيَّةُ  |
| `eu-ES`       | Basque                | euskara          |
| `bg-BG`       | Bulgarian             | Български        |
| `ca-ES`       | Catalan               | català           |
| `zh-Hans`     | Chinese (Simplified)  | 中文(简体)       |
| `zh-Hant`     | Chinese (Traditional) | 中文(繁體)       |
| `hr-HR`       | Croatian              | hrvatski         |
| `cs-CZ`       | Czech                 | čeština          |
| `da-DK`       | Danish                | dansk            |
| `nl-NL`       | Dutch                 | Nederlands       |
| `en-US`       | English               | English          |
| `et-EE`       | Estonian              | eesti            |
| `fi-FI`       | Finnish               | suomi            |
| `fr-FR`       | French                | français         |
| `gl-ES`       | Galician              | galego           |
| `de-DE`       | German                | Deutsch          |
| `el-GR`       | Greek                 | Ελληνικά         |

> | `he-IL`       | Hebrew                | עברית            |

Reviewer: Clean up table spacing

Author: Actually it's clean, but some letters are not monospace, thus, it looks weird.

| `pl-PL`       | Polish                | Polski           |
| `pt-BR`       | Portuguese (Brazil)   | Português        |
| `pt-PT`       | Portuguese (Portugal) | português        |
| `ro-RO`       | Romanian              | română           |
| `ru-RU`       | Russian               | Русский          |
| `sr-Cyrl`     | Serbian (Cyrillic)    | српски           |
| `sr-Latn`     | Serbian (Latin)       | srpski           |
| `sk-SK`       | Slovak                | slovenčina       |
| `sl-SI`       | Slovenian             | slovenski        |
| `es-ES`       | Spanish               | español          |
| `sv-SE`       | Swedish               | svenska          |
| `th-TH`       | Thai                  | ไทย              |
| `tr-TR`       | Turkish               | Türkçe           |
| `uk-UA`       | Ukrainian             | українська       |
| `vi-VN`       | Vietnamese            | Tiếng Việt       |

### Community-contributed languages

| Language Code | Language        | Name           | Translator                               |
| ------------- | --------------- | -------------- | ---------------------------------------- |

> | `ar-EG`       | Egyptian Arabic | اللهجه المصريه | [@midineo](https://github.com/midineo)   |

Reviewer: add `ar-JO`

Reviewer: table cleanup

| `zh-HK`       | Cantonese       | 廣東話         | [@compulim](https://github.com/compulim) |

### JSON file format

> This file format is maintained by localization team at Microsoft. An automated pipeline has been set up and will help Web Chat to translate new strings in a matter of days than weeks or months.

Reviewer: ```suggestion
This file format is maintained by the localization team at Microsoft,. where an automated pipeline allows for Web Chat strings to be translated in a timely manner.
```

### Lock down on strings

> For strings that are maintained by Microsoft, they have passed additional checks, for example, profanity check and cultural check, to guarantee the quality of the strings. In order to maintain the quality of these strings, modification of these strings requires additional checks from the localization team.

Reviewer: ```suggestion
Localized strings maintained by Microsoft have passed additional checks, including profanity, and cultural checks to guarantee the localization quality. For maintenance purposes, any requested changes must be validated by the localization team.
```

> The additional checks do not applies for strings that are contributed by the community, for example, Egyptian Arabic and Cantonese.

Reviewer: ```suggestion
Additional validation is not required for languages that are contributed by the community, e.g. Egyptian Arabic and Cantonese.
```

### Overrideable by developers

> Developers should have a way to modify existing strings. One of the popular ask is the modify the "Type your message" placeholder text in the message input box.

Reviewer: ```suggestion
Developers may modify existing strings for customization purposes. One popular ask is the modification of the "Type your message" placeholder text in the message input box.
```
message input box => SendBox?

Author: I prefer to keep it as message input box, because in short future, we will have speech input box, and other input box. I can also name it "text input box".

What do you think?

> ### New hooks to simplify internationalization

Reviewer: ```suggestion
### Hooks to streamline internationalization
```

> `useXXXFormatter` hooks are added to simplify internationalization using Unicode CLDR data

Reviewer: ```suggestion
All `useXXXFormatter` hooks are intended for internationalization using Unicode CLDR data
```

> -  `useByteFormatter` will format number of file size in bytes, into localized strings

Reviewer: ```suggestion
-  `useByteFormatter` will format file size-related numbers into localized strings
```

> -  `useDateFOrmatter` will format date in absolute format, similar to `useLocalizeDate` but with updated function signature. This hook do not use Unicode CLDR data

Reviewer: ```suggestion
-  `useDateFOrmatter` will format date in absolute format, similar to `useLocalizeDate`, but with an updated function signature. This hook does not use Unicode CLDR data
```

> -  `useRelativeTimeFormatter` will format time in relative format, using Unicode CLDR data

Reviewer: ```suggestion
-  `useRelativeTimeFormatter` will format time to relative format, using Unicode CLDR data
```

> -  `useLanguage` now accept options to return oral language instead of written language

Reviewer: ```suggestion
-  `useLanguage` now accepts options to return oral language instead of written language
```

## Appendix

### Convert string ID to new format

> To support the work from localization team, we are refreshing all string ID in a standard format.

Reviewer: ```suggestion
To support the localization team, all string IDs are being refreshed to a standard format.
```

> The following code is used by Web Chat team to convert strings from our previous format into new format.

Reviewer: ```suggestion
The following code is used by the Web Chat team to convert strings from our previous format into new format.
```

For strings that are maintained by Microsoft, please file a bug to the Web Chat repo. Validation will be performed on these changes to maintain high quality.

For strings that are contributed by the community, please submit a pull request to the Web Chat repo.

## Adding new strings for new UI

To add new strings to be used in the UI, please update `/packages/component/src/localization/en-US.json`. The localization team will pick up the new strings and translate them to all other officially supported languages.

## Overriding localization strings

Developers can override localization strings in one of two ways. An `overrideLocalizedStrings` prop can be passed either as a plain object or a function in the Web Chat renderer.

```js
window.WebChat.renderWebChat(
   {
      locale: 'en-US',

      overrideLocalizedStrings: (strings, language) => ({
         ...strings,
         TEXT_INPUT_PLACEHOLDER: 'What is on your mind?'
      }),

      overrideLocalizedStrings: {
         TEXT_INPUT_PLACEHOLDER: 'What is on your mind?'
      }
   },
   document.getElementById('webchat')
);
```

## Design

### User stories

This localization feature will support the following user stories:

> -  Our strings will be automatically send thru a pipeline to Microsoft localization team, which they will translate the string and send us the pull request

Reviewer: ```suggestion
-  Web Chat UI strings will be automatically sent through a pipeline to the Microsoft localization team, which they will translate and perform a pull request for
```

   -  The pipeline is implemented and owned by the localization team
-  Strings localized by Microsoft localization team will be locked down

>    -  Web Chat team can recommend more suitable strings to localization team, it will go through validation and send back as a pull request

Reviewer: ```suggestion
   -  Web Chat team can recommend more suitable strings to localization team; the request will be reviewed and validated, then sent back as a pull request
```

> -  Developers can override strings for their own setup of Web Chat, without forking our code

Reviewer: ```suggestion
-  Developers can override strings for their own setup of Web Chat, without forking our code
```
```suggestion
-  Developers can override strings for their own setup of Web Chat without forking our code
```
